### PR TITLE
implement DList_ForEach

### PR DESCRIPTION
**Pull request by parth21999**

### CODE REVIEW

File: src/doublylinkedlist.c [hunk 1]
```diff
@@ -110,8 +113,47 @@ void DList_InsertHeadList(PDLIST_ENTRY listHead, PDLIST_ENTRY entry)
...
+    else
+    {
+        PDLIST_ENTRY entry = listHead->Flink;
+        while (entry != listHead)
+        {
+            bool continueProcessing;
+            /*Codes_SRS_DLIST_43_009: [DList_ForEach shall call actionFunction on each entry in the list defined by listHead along with actionContext.]*/
+            if (actionFunction(entry, actionContext, &continueProcessing) != 0)
+            {
+                LogError("failure in actionFunction(entry, actionContext, &continueProcessing)");
```
Comment (anporumb) [inline src/doublylinkedlist.c:137]: >entry, actionContext, &continueProcessing [](http://example.com/codeflow?start=52&length=41)

parameteres to actionFunction should be logged as values too.

There's never enough logging in the world of services.

File: src/doublylinkedlist.c [hunk 2]
```diff
@@ -110,8 +113,47 @@ void DList_InsertHeadList(PDLIST_ENTRY listHead, PDLIST_ENTRY entry)
 
 IMPLEMENT_MOCKABLE_FUNCTION(, int, DList_ForEach, PDLIST_ENTRY, listHead, DLIST_ACTION_FUNCTION, actionFunction, void*, actionContext)
 {
-    (void)listHead;
-    (void)actionFunction;
-    (void)actionContext;
-    return 0;
+    int result = 0;
```
Comment (anporumb) [inline src/doublylinkedlist.c:116]: >= 0; [](http://example.com/codeflow?start=15&length=4)

function result should not be initialized here.

File: tests/doublylinkedlist_ut/doublylinkedlist_ut.c [hunk 3]
```diff
@@ -19,17 +24,62 @@ static simpleItem simp5 = { 5, { NULL, NULL } };
...
+    ASSERT_ARE_EQUAL(int, 0, umock_c_init(on_umock_c_error), "umock_c_init");
+    ASSERT_ARE_EQUAL(int, 0, umocktypes_charptr_register_types(), "umocktypes_charptr_register_types");
+
+    REGISTER_UMOCK_ALIAS_TYPE(PDLIST_ENTRY, void*);
+
 }
 
 TEST_SUITE_CLEANUP(TestClassCleanup)
 {
+    umock_c_deinit();
```
Comment (anporumb) [inline tests/doublylinkedlist_ut/doublylinkedlist_ut.c:80]: >umock_c_deinit(); [](http://example.com/codeflow?start=4&length=17)

How did you find this is missing?

File: tests/doublylinkedlist_ut/doublylinkedlist_ut.c [hunk 4]
```diff
@@ -414,5 +465,122 @@ TEST_FUNCTION_CLEANUP(TestMethodCleanup)
...
+        ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+        ASSERT_ARE_NOT_EQUAL(int, 0, result);
+    }
+
+    /*Tests_SRS_DLIST_43_009 : [DList_ForEach shall call actionFunction on each entry in the list defined by listHead along with actionContext.]*/
+    /*Tests_SRS_DLIST_43_011 : [DList_ForEach shall succeed and return zero.]*/
+    TEST_FUNCTION(DList_ForEach_succeeds)
+    {
+        ///arrange
+        DLIST_ENTRY head = { 0 };
```
Comment (anporumb) [inline tests/doublylinkedlist_ut/doublylinkedlist_ut.c:501]: >= { 0 }; [](http://example.com/codeflow?start=25&length=8)

is this initialization necessary?

File: tests/doublylinkedlist_ut/doublylinkedlist_ut.c [hunk 5]
```diff
@@ -414,5 +465,122 @@ TEST_FUNCTION_CLEANUP(TestMethodCleanup)
...
+        DList_InitializeListHead(&head);
+        DList_InsertTailList(&head, &(simp1.link));
```
Comment (anporumb) [inline tests/doublylinkedlist_ut/doublylinkedlist_ut.c:503]: >Tail [](http://example.com/codeflow?start=20&length=4)

I've always found funny how the fields that hold the dlist are called flink/blink, but we use Tail And Head.

InsetAtFlink :)

./disconsider this comment of mine.

File: tests/doublylinkedlist_ut/doublylinkedlist_ut.c (hunk 4 above)
Comment (anporumb) [reply tests/doublylinkedlist_ut/doublylinkedlist_ut.c:501]: Here and everywhere

File: tests/doublylinkedlist_ut/doublylinkedlist_ut.c (hunk 5 above)
Comment (anporumb) [reply tests/doublylinkedlist_ut/doublylinkedlist_ut.c:503]: Head or tail is for coin tosses

**Review by anporumb — APPROVED** [review]
:shipit:

File: tests/doublylinkedlist_ut/doublylinkedlist_ut.c (hunk 4 above)
Comment (parth21999) [reply tests/doublylinkedlist_ut/doublylinkedlist_ut.c:501]: Not necessary, removing.

File: tests/doublylinkedlist_ut/doublylinkedlist_ut.c (hunk 3 above)
Comment (parth21999) [reply tests/doublylinkedlist_ut/doublylinkedlist_ut.c:80]: vld

File: src/doublylinkedlist.c (hunk 2 above)
Comment (parth21999) [reply src/doublylinkedlist.c]: Removed.

**Pull request merged**